React/Views/RCTWebView.m

```diff
@@ -208,13 +208,68 @@ - (BOOL)webView:(__unused UIWebView *)webView shouldStartLoadWithRequest:(NSURLR
     };
   });
 
+  NSString *sourceURL = (NSString *)self.source[@"uri"];
+  NSString *requestURL = (request.URL).absoluteString;
+  NSString *namedNavigationType = navigationTypes[@(navigationType)];
+
+  NSMutableDictionary<NSString *, id> *event = [self baseEvent];
+  [event addEntriesFromDictionary: @{
+                                     @"url": requestURL,
+                                     @"navigationType": namedNavigationType
+                                     }];
+
+  BOOL isJSNavigation = [request.URL.scheme isEqualToString:RCTJSNavigationScheme];
+  BOOL isTopFrame = [request.URL isEqual:request.mainDocumentURL];
+  BOOL isLoadingSourceURL = ([sourceURL isEqualToString:requestURL] && !_webView.isLoading);
+  
+  // skip this for the JS Navigation handler, initial load, iFrames
+  if (!isJSNavigation && isTopFrame && !isLoadingSourceURL) {
+    if (_navigationBlockingPolicies.count > 0) {
+      NSDictionary *currentValues = @{
+                                      @"currentURL": sourceURL,
+                                      @"url": requestURL,
+                                      @"navigationType": namedNavigationType
+                                      };
+      NSString *eventValue;
+      BOOL policyFulfilled;
+
+      for (NSDictionary *policy in _navigationBlockingPolicies) {
+        // policy with no rules
+        if (policy.count < 1) {
+          continue;
+        }
+
+        policyFulfilled = YES;
+
+        for (NSString *key in policy) {
+          // policy has failed already
+          if (!policyFulfilled) break;
+
+          eventValue = [currentValues objectForKey:key];
+
+          // unverifiable policy rule
+          if (!eventValue) {
+            RCTLogWarn(@"Could not verify webview loading policy named %@. Failing policy with rules %@", key, policy);
+            policyFulfilled = NO;
+            break;
+          }
+
+          NSPredicate *rule = [NSPredicate predicateWithFormat:@"SELF MATCHES %@", [policy objectForKey:key]];
+
+          policyFulfilled = [rule evaluateWithObject: eventValue];
+        }
+
+        // call blocked callback
+        if (policyFulfilled && _onNavigationBlocked) {
+          _onNavigationBlocked(event);
+          return NO;
+        }
+      }
+    }
+  }
```

Comment:
This seemed to be the way it is done in this class, and similar ones. E.g.: take a look at the `webViewDidFinishLoad` implementation. I agree with you about readability, just want to make sure this is the way to go.